CPT: Include all available post actions as independent menu item components

aduth opened the pull request:
This pull request seeks to implement most all post actions to be available for custom post types and adds previously unaccounted-for capabilities checking and labels as derived from the post type object.

Specifically, actions include: "View", "Publish", "Edit", "Restore", "Trash" / "Permanently Delete"

![Actions](https://cloud.githubusercontent.com/assets/1779930/16469104/373213c2-3e1d-11e6-8d2a-3d5d0728e80f.png)

__Implementation notes:__

As I started fleshing out `<PostTypeListActions />` with individual action buttons, I found that the component became quite large with logic and props. It therefore made sense to me to instead allow the action menu items to be independent, with their own rendering behavior and click handling.

`<PopoverMenuItem />` has been enhanced to allow rendering as an anchor tag (link), and to accept props not otherwise listed in the `propTypes` (e.g. `target`).

__Testing instructions:__

Verify that post actions on the [custom post types list screen](http://calypso.localhost:3000/types/post) are consistent with the status and type of the post, and that actions behave as you'd expect (with the exception of Restore, which is to be implemented in a future pull request). Verify that actions are not made available if your user account does not have permission to perform that action.

Test live: https://calypso.live/?branch=add/cpt-actions

mtias commented:
Could we hook up the "View Project" with WebPreview?

mtias commented on client/my-sites/post-type-list/post-type-list-post-actions/edit.jsx:
```diff
@@ -0,0 +1,61 @@
+/**
```
I'd be ok with calling this one `ellipsis-post-actions`.

aduth replied:
>I'd be ok with calling this one `ellipsis-post-actions` or `post-actions-ellipsis`.

Yeah, could probably do for a rename. I'd probably prefer `menu` in the name somewhere, either `post-actions-menu` or `post-actions-ellipsis-menu` ?

mtias commented on client/my-sites/post-type-list/index.jsx:
```diff
@@ -153,6 +154,7 @@ class PostTypeList extends Component {
 						siteId={ siteId }
 						query={ { ...query, page } } />
 				) ) }
+				{ siteId && <QueryPostTypes siteId={ siteId } /> }
```
Why do you need this one? It seems it would make reusing this less flexible as it'll trigger post types requests on its own.

aduth replied:
>Why do you need this one? It seems it would make reusing this less flexible as it'll trigger post types requests on its own.

See extended description of 03ea849.

mtias replied:
Why is that so and how is it bad?

aduth replied:
>Why is that so and how is it bad?

Because requests are triggered by the mount. I assume `<PopoverMenu />` is only rendering the menu items when the menu is visible. I'd thought to move it to the base menu component, but this is even worse in that requests are triggered both after the initial request of posts and when VirtualScroll changes the rendered indices.

Whether it's "bad" that types are requested on each visible toggle... maybe not. Could potentially be a bad user experience if they've not yet been loaded (but it's likely they have been, since My SItes sidebar requires the types). Likely quite a bit of unnecessary requesting, though.

Almost wonder if we should drop the post type specific labels, since they're already a bit inconsistent (there's no "Trash Project" or "Publish Project" labels available). "View", "Edit" could serve just as well across any post type? The Edit component also pulls capability from the post type, but I'm a bit worried about this anyways, since our site endpoint doesn't return _all_ of the capabilities (nothing to stop someone from setting the edit capability for their custom post type to `switch_themes`).

aduth replied:
Alternatively, while it goes against our "keeping fresh" methodology of query component mounting, we could render the `<QueryPostTypes />` within the menu items, but only if types aren't yet known for the site.

mtias replied:
Regarding labels I agree simple verbs would be better.

Regarding the queries, it seems multiple queries should have a way to not run if there's another that was already run by a parent within a certain interval. But it still seems that QueryPostTypes is something I would run in my `main.jsx` component, not on the list (which would run the QueryPosts).

aduth replied:
>Regarding the queries, it seems multiple queries should have a way to not run if there's another that was already run by a parent within a certain interval.

It will consolidate if there are multiple components mounting `<QueryPostTypes />` at the same time (or for that matter, if a request is in progress at all), from [`requestingPostTypes` checks in the component](https://github.com/Automattic/wp-calypso/blob/178d0639c1ba80384596ce9f2868ae3ebdde5f89/client/components/data/query-post-types/index.jsx#L22).

>But it still seems that `QueryPostTypes` is something I would run in my `main.jsx` component, not on the list (which would run the `QueryPosts`).

I'd probably disagree with this point in that it moves us back towards our original problems of ancestor components needing to be concerned with the data requirements of their descendants (rather than the components being self-sufficient) and the burden on the developer to remember to do so.

mtias commented on client/my-sites/post-type-list/post-type-list-post-actions/index.jsx:
```diff
@@ -0,0 +1,32 @@
+	return (
+		<div className="post-type-list-post-actions">
+			<EllipsisMenu position="bottom left">
+				<PostTypeListPostActionsView globalId={ globalId } />
```
How would you handle custom actions? Maybe this can be constructed outside `post-type-list-post-actions` using the individual `import PostTypeListPostActionsPublish from './publish';`

aduth replied:
>How would you handle custom actions?

I thought about this a bit. In terms of customization, do we want to enable a developer to replace the entire set of actions, or just append new custom options?

In either case, maybe this component could accept a prop `actions`, an array of menu item components, defaulting to the ones listed here, each being passed `globalId` of the post.

mtias replied:
My experience with dropdown-menu is that arrays become inflexible and just declaratively setting the children is preferred. I'd be ok with making any children appear after an hr division and treating the default set as static. Alternatively we could have boolean props for the base four actions that you could disable if needed.

aduth replied:
Could be a single boolean prop `includeDefaultActions={ false }`, and the developer could reincorporate any they'd want to keep in children?

```js
<PostTypeListPostActions includeDefaultActions={ false }>
	<PostTypeListPostActionsEdit />
	<MyCustomAction />
</PostTypeListPostActions>
```

mtias replied:
That seems sensible, as long as I can add children to the base set as well without setting it to false. The pages popover menu is a good example to keep in mind.

aduth replied:
Right, I'd think the following would include the default set of actions, a separator, then my custom action:

```js
<PostTypeListPostActions>
    <MyCustomAction />
</PostTypeListPostActions>
```

aduth commented:
>Could we hook up the "View Project" with WebPreview?

Sounds like a good idea, but will probably save it for a subsequent pull request.

mtias commented:
> Sounds like a good idea, but will probably save it for a subsequent pull request.

Yes, just wanted to mention in case we avoid adding the extra props for href and target to menu item.

aduth commented:
>Yes, just wanted to mention in case we avoid adding the extra props for href and target to menu item.

Ah, fair point. I could change this to a button stub with "Not Yet Implemented" behavior as I've done with a few of the others, if we'd prefer to avoid adding support for these to menu item.

aduth commented:
Pushed up a handful of changes addressing all feedback, particularly:

- Renaming menu component to `<PostActionsEllipsisMenu />` (4b8fa2f)
- Rendering `<QueryPostTypes />` in single component where it's used only if not already known (311baa8)
- Checking validity of capability of custom post type (i.e. available on site response) before using it (d16b177)
- Allow customization of action menu items through `children` and/or `includeDefaultProps` (default: `true`, f14230e)

Kept `<PopoverMenuItem />` `href` enhancements because while we'll want Preview to open in `<WebPreview />`, we'll still need the "Edit" button to link to the post editor.
